File: utils/cpp/src/Comm/Hub.cpp
#include <Comm/Hub.hpp>

template <class T>
Comm::Hub<T>::Hub(Bridge<T>* bridge)
 : bridge(bridge) { }

template <class T>
void Comm::Hub<T>::emit(T name, std::vector<T> message){
  message.push_back(name);
  this->bridge->send(message);
}

template <class T>
void Comm::Hub<T>::on(T name, handlerFunction<T> handler){
  this->handlers[name].push_back(handler);
}

template <class T>
void Comm::Hub<T>::poll(){
  std::queue<std::vector<T>> messages = this->bridge->receive();
  while(!messages.empty()){
    std::vector<T> message = messages.front();
    std::vector<handlerFunction<T>> handlers = this->handlers[message.back()];
    message.erase(message.end() - 1);
    for(unsigned int i = 0; i < handlers.size(); ++i){
      handlers[i](message);
    }
    messages.pop();
  }
}

File: utils/javascript/src/comm/hub.js
const hub = bridge => {
  const handlers = {};
  const hub = {
    on: (name, handler) => {
      if(!handlers[name]){
        handlers[name] = []
      }
      handlers[name].push(handler)
    },
    poll: () => {
      const messages = bridge.receive()
      messages.forEach(message => {
        const name = message[0]
        message.splice(0, 1)
        if(handlers[name]){
          handlers[name].forEach(handler => handler(hub, message))
        }else{
          console.log('Received event with no listener', name, message)
        }
      })
    },
    emit: (name, data) => {
      bridge.send([name].concat(data))
    }
  }
  return hub
}

module.exports = hub
